### ISSUE TITLE
Button within GUI for starting the scraping

### ISSUE THREAD

**Issue by Aidolii** [fake-name/xA-Scraper#109]
I'm trying to set up a portable Windows environment for xA-Scraper, but the part I'm stuck at has to do with the console. Namely, the scraper would have to be running in the background within the session for me to send the appropriate console command.

Can you just add a button within the GUI that does it?

**Comment by Aidolii** [comment]
Ok, well for some reason I can't get it to work with my setup. I don't know why. It would still be very nice to have a button within the GUI that can do it.

**Comment by fake-name** [comment]
> Ok, well for some reason I can't get it to work with my setup.

Define "can't get it to work". What happens when you run `python3 -m manage`?

**Comment by Aidolii** [comment]
I just get 'No module named manage'

But even if I can get it to work, I would still really appreciate a GUI button for it.

**Comment by fake-name** [comment]
You need to be in the root directory of the project for that (the same folder with the `main.py` file).

**Comment by Aidolii** [comment]
Work has kept me busy so it took me a bit to follow up on this. As I expected, it does not work as you say it does:

```
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

What it describes is not really possible to do in a Windows environment. I need to start up a second command prompt within the same instance to run the command through.

**Comment by fake-name** [comment]
1. You don't need anything running. It doesn't tell a running scraper to run, it actually executes locally.
2. I have *no* idea what's going on. The only place I've seen `freeze_support()` is when building packed executables.
3. I don't do any testing on windows.

**Comment by Aidolii** [comment]
Well shit.

**Comment by Aidolii** [comment]
Is there anything I can do or am I just stuck? Because I don't know what to do either. I get the error whether or not I'm running the server, so I figured it was instance based. If it runs independently. then... I haven't a clue what to do.

**Comment by fake-name** [comment]
![fetchinating 0](https://user-images.githubusercontent.com/1401239/39913318-48281800-54b6-11e8-905f-089cd91a434e.gif)

I thought I had a note in the readme about that, but I guess it was another project.

Is there a more complete traceback?

**Comment by Aidolii** [comment]
Okay, so it seems that for some reason, if I do a site individually, it works fine, but fetch-all throws errors.

Unfortunately in doing so I ran into a new problem. It seems 'fetch fa' leads to a ton of 503 errors, perhaps the data is being access too fast or it's recognizing whatever user agent it's using as a bot. Dunno what to do about it.

**Comment by fake-name** [comment]
CuddleBear92 apparently has had success running the scraper under the [WSL](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux), so if you're on windows 10, that might be an option.

**Comment by fake-name** [comment]
Ohhh, yeah, `fetch-all` can be a little bit *to* aggressive.  `python3 -m manage fetch` (without arguments) fetches every plugin, but in sequence.

**Comment by Aidolii** [comment]
I see, I figured fetch would be for individual plugins only and fetch-all would be all plugins.

Anyway, the 503 problem, it does start to retrieve some stuff, but then it starts hitting that 503 and it's like a trainwreck. Any ideas on how to fix it?

**Comment by fake-name** [comment]
> fetch [sitename]
>        **with no sitename, this executes all plugins in sequence.**
>        With a sitename, executes the named plugin.

WRT the 503 issues, I see them a lot because of removed art, or artist's that have banned/removed. Can you check if you can access the URL that produces the 503 in an actual browser?

**Comment by Aidolii** [comment]
Yes, I can access it in a browser just fine. It's in fact throwing a new 503 error per page, pages that are all still up. I've scraped this artist with a chrome extension in the past and it went just fine.

**Comment by fake-name** [comment]
I haven't got any idea what's going on, then. I haven't seen that particular problem in any of my testing.

**Comment by Aidolii** [comment]
I'm pretty sure it's FA throttling it because it's detecting it's a bot.

On your end, can you try adding an artist with a lot of art that you don't already have added, and then running the plugin? It's possible you aren't running into the problem because you haven't had to scrape a new artist on fa for a while and you're only grabbing new submissions, which isn't enough to trigger the detection.

**Comment by fake-name** [comment]
Oh derp, you said FA, not DA.

**Comment by Aidolii** [comment]
This extension never hits a throttling problem for whatever reason, you might be able to bust it open and look at the code.

https://chrome.google.com/webstore/detail/nmcffcaebiikdmiklndccpkagmllckkh

**Comment by fake-name** [comment]
Yep, that's based on reversing JS (which I've done, but isn't fun).

Mostly, I can't fix things I haven't seen, and I haven't seen this particular failure mode (and I've just scraped ~100 MB of new content, and still going).

**Comment by Aidolii** [comment]
Do you have an email? I feel uncomfortable naming specific artists and sharing logs that contain paths that I don't want appearing on a google search.

**Comment by fake-name** [comment]
<anything> @ fake-url [dot] com (it's a catchall address)

**Comment by fake-name** [comment]
<sub>That was supposed to be sarcastic</sub>

**Comment by Aidolii** [comment]
Sorry. Just trying to do a lot of things at once atm.

**Comment by fake-name** [comment]
Lelz, no problem.

**Comment by Aidolii** [comment]
Let me know if the log identifies the problem, I'd love to have this working.

**Comment by fake-name** [comment]
I'll dig about tomorrow. ATM I'm going to sleep.

**Comment by Aidolii** [comment]
Alright, take care.

**Comment by fake-name** [comment]
Ok, so I ran into this issue at some point last night.

It looks like they have some simple request-rate based limiting. I added some sleeps to the scraper, and it seems to have ameliorated the issue. 

It's still running, so whether this actually fixed the issues remains to be seen.

**Comment by Aidolii** [comment]
Well now today I start it up with the same script I used last night and I hit this error.

```
Setting up loggers....
done
Setup
initialized manager
fetch args ['fa'] <class 'list'>
Cannot find plugin for name <class 'rewrite.modules.fa.faScrape.GetFA'>!

Missing run command: None
```

It wasn't doing this last night...

**Comment by fake-name** [comment]
Hahahaha, wow I'm dumb. I fiddled with how the multiprocessing startup call is serialized, and apparently broke that. DERP.

Hangon.

**Comment by fake-name** [comment]
Should work again.

Also, I think `fetch-all` might be functional too.

**Comment by Aidolii** [comment]
Okay it seemed like it was frozen for a WHILE and at some point it started downloading, so all is good.

**Comment by fake-name** [comment]
> Okay it seemed like it was frozen for a WHILE

It was probably flushing to the database.

The insertion behaviour is optimized for a context where it's adding a few items to already existing set. and particularly if you're using sqlite, you'll probably wind up wating for a BUNCH of `fsync()` operations.

**Comment by Aidolii** [comment]
Yeah, I am using sqlite, don't really know how to work MySQL.

The sleep times probably arent' nearly as necessary for downloading as they are for link gathering. In fact I bet it would work just fine if there was no sleep time at all for downloading. With link gathering it's making a ton of requests in short succession, whereas downloading is going to be slower.

The best choice would probably be to measure it out, like 'Did this take more than 2 seconds? If so, do not sleep. Did it take less than 2 seconds? If so, sleep until it would have been 2 seconds.' (2 seconds is just an example.)

**Comment by Aidolii** [comment]
What would happen if I used that extension I linked earlier and stuck the file output where it would normally download, would it recognize that I have them all already, or would it try to download and overwrite them?

If it can but has to do so through HTTP 304 Not Modified, would the new sleep you put in make it so it has to sleep every single time?

**Comment by fake-name** [comment]
> don't really know how to work MySQL.

That's fine, the scraper doesn't work with mysql either. 

PostgreSQL, on the other hand, works fine.

> The best choice would probably be to measure it out, like 'Did this take more than 2 seconds? If so, do not sleep. Did it take less than 2 seconds? If so, sleep until it would have been 2 seconds.' (2 seconds is just an example.)

I specifically hit limiting while fetching images, so I think it can happen on both.

> What would happen if I used that extension I linked earlier and stuck the file output where it would normally download, would it recognize that I have them all already, or would it try to download and overwrite them?

Probably not. It does skip things if it finds the file already exists, but unless the other scraper uses the **exact same** naming scheme, it won't help.

**Comment by Aidolii** [comment]
>Probably not. It does skip things if it finds the file already exists, but unless the other scraper uses the exact same naming scheme, it won't help.

It looks like your scraper doesn't rename the files at all though. Neither does the extension.

**Comment by fake-name** [comment]
Give it a try, then?

**Comment by Aidolii** [comment]
I will when this finishes. Which is probably going to be about... 3-1/2 hours. 

Does interrupting it screw things up?

**Comment by fake-name** [comment]
It shouldn't. It'll have to rewalk the gallery listings, but other then that it's harmless.

You may need to run `python3 -m manage reset-run-state` to reset the run-state flags in the database. Basically, it tracks what plugins are running, to prevent the same plugin from running multiple times.

**Comment by Aidolii** [comment]
Would the last file it was downloading show erroneously as complete when it's actually only a partial download?

**Comment by fake-name** [comment]
Files should only be marked as complete when they've been fully written to disk.

e.g.

1. Start Downloading
2. Finish Downloading
3. Write File to disk
4. Mark as downloaded

------

I should add that the design for the scraper is to be run on a always-on computer, so it's not optimized for scraping speed. Basically, I have a couple servers in my closet that are always on.

**Comment by Aidolii** [comment]
It seems to have started over completely, and it's just overwriting the files I copied into there. I lost my 750/2000 progress from interrupting it.

Urgh...

**Comment by fake-name** [comment]
It should skip files fetched in the previous execution, but order may not be the same (the order it tries to retrieve things is not deterministic).

**Comment by Aidolii** [comment]
Ok, it was successful, but when I start up the GUI, things are rather strange. Every single picture is just a broken thumbnail, and the 'add time' is strange, like it says some pics were added 73 days ago, and then another set of 144, 355, etc. Just weird numbers that seem random. 

The [src] and Image Source URL columns are also wrong. The Image Source URLs should be in the [src] hotlink and the Image Source URLs should match the file that was downloaded, e.g. http://d.facdn.net/art/artist/1234567890/1234567890.artist_2018-05-11.png

**Comment by Aidolii** [comment]
They're also all completely out of order. There's no date sorting or even sorting by name.

**Comment by Aidolii** [comment]
I'm looking at the console and I'm noticing that for some reason, it's looking in the rewrite directory for 'absolute_path_downloads_will_go_here', but that's in the root directory.

**Comment by fake-name** [comment]
`[src]` goes to the artist's page (it means artist source). 

Images work for me, but if you're saying it's looking in the rewrite directory, it sounds like the `absolute_path_downloads_will_go_here` isn't absolute.

The download path is built as: `settings['dldCtntPath'], img_row.fspath)`, where `img_row.fspath` should be a path relative to `settings['dldCtntPath']`. However, the whole affair will be calculated as relative to the CWD of the python interpreter if you don't have an absolute path in `settings['dldCtntPath']`.

I have no idea what odd effects a non-absolute path will have.

The add time is when the item was added *to FA*, not when the item was retreived.

Sort order is hard-coded to item upload date, in descending order. It's not currently changeable.

The entire webshit was kind of bolted on.

**Comment by fake-name** [comment]
> I dunno why you'd need source to point to the [src] when the artist's name already does that.

I can't remember the specific circumstances, but I think I was dealing with some issues related to proper artist linking, and I added it to make debugging easier.

The releases are explicitly ordered by the addtime

```
	releases = db.session.query(database.ArtItem) \
		.filter(database.ArtItem.state == "complete") \
		.order_by(desc(database.ArtItem.addtime)) \               <-----------
		.options(joinedload('artist'))     \
		.options(joinedload('files'))      \
		.options(joinedload('tags'))
```

It *should* be sorted, unless sqlite is doing something stupid.

**Comment by Aidolii** [comment]
Oh, it most definitely is SQLite doing something stupid. Do you know of an easy, python-only solution to getting a PostgreSQL database up and running locally, that can be called on-demand?

**Comment by fake-name** [comment]
Python is basically irrelevant to PostgreSQL. 

You'd have to install it yourself, and then getting the users and such set up is kind of fiddly.

**Comment by fake-name** [comment]
Ok, the sort order stuff is because sqlite [doesn't have a `datetime` type](http://docs.sqlalchemy.org/en/latest/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME).  

As such, the date/time is stored as a string, and then sorted lexicographically.

Looking at the schema used, I'd *think* it would still sort correctly, but I'm not sure what actual SQL SQLAlchemy is emitting for that case.

**Comment by Aidolii** [comment]
I can assure you it's not sorting correctly. But honestly there's so many problems with the GUI I'm not sweating it. It doesn't seem to have the right filepaths in the interface for some reason, either.

**Comment by fake-name** [comment]
Oh, I wasn't doubting you, just confused.
